File: cpp/Hash_Table/0001_Two_Sum.cpp
#include <vector>
#include <unordered_map>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  vector<int> twoSum(vector<int> &nums, int target) {
    unordered_map<int, int> map;
    for (int i = 0; i < nums.size(); ++i) {
      if (map.find(target - nums[i]) != map.end()) return {i, map[target - nums[i]]};
      map[nums[i]] = i;
    }
    return {};
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0001-two-sum.py
class Solution:
    def twoSum(self, nums: [int], target: int) -> [int]:
        record = dict()
        for idx, val in enumerate(nums):
            if target - val not in record:
                record[val] = idx
            else:
                return [record[target - val], idx]


nums = [2, 7, 11, 15]
target = 9
print(Solution().twoSum(nums, target))  # [0,1]

nums = [3, 2, 4]
target = 6
print(Solution().twoSum(nums, target))  # [1,2]

nums = [3, 3]
target = 6
print(Solution().twoSum(nums, target))  # [0,1]
